### PR TITLE
[Metricbeat] Discard partial Logstash pipeline documents

### PR DESCRIPTION
**Pull request by Kerry350**
## What does this PR do?

This PR closes https://github.com/elastic/kibana/issues/133230.

This PR is the Metricbeat module side of a three pronged fix to a race condition between Metricbeat and the Logstash APIs, which leads to partial data being ingested.

I've also updated the `710` fixture, as it had data from `7.8.1`.

It looks like these files haven't been linted before as there are a lot of linting errors and the linter is (AFAIK) pretty new. I haven't fixed these as they are heavily unrelated and belong to, mostly, only two main groups of issues.

## Why is it important?

This is important as this partial ingestion is not expected. There *should* always be a `hash` assigned (at least after version 7.3.0). This leads to the Stack Monitoring UI breaking as we can't rely on the data containing the properties needed.

## Checklist

- [x] My code follows the style guidelines of this project (I think so, couldn't find a STYLEGUIDE file)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Note: I have also left a couple of inline code / review comments.

- Follow the docs to setup your environment to build Beats (here: https://www.elastic.co/guide/en/beats/devguide/current/beats-contributing.html#setting-up-dev-environment)

- Checkout this branch, and then`cd metricbeat`, and then `mage build`.

- Make a Metricbeat config file that contains the following (change the hosts and credentials as needed, this is an example). The `period: 1s` is important so we can capture the race condition:

```yml
http.enabled: true
metricbeat.modules:
  - module: system

  - module: logstash
    xpack.enabled: true
    period: 1s
    hosts: [ "localhost:9600" ]

output.elasticsearch:
  hosts: [ "localhost:9200" ]
  username: "elastic"
  password: "changeme"
```

- Then run `./metricbeat -e -c PATH_TO_YOUR_METRICBEAT_YML_FILE`

- Run Logstash. Docker works reliably in my testing with the `1s` period.

```
docker run --name logstash \                
--pull always --rm \
--hostname=logstash \
--publish=9600:9600 \
--volume="<PATH TO YOUR LOGSTASH YML>:/usr/share/logstash/config/logstash.yml:ro" \
docker.elastic.co/logstash/logstash:master-SNAPSHOT \
-e 'input { java_generator { eps => 1 } } output { file { path => "/dev/null" } }'
```

- The Beat should now poll at a time when a partial result is returned from the Logstash API. It should then discard this document, and it should not be ingested. You can verify this in the logs.

Screenshot 2022-06-12 at 00 51 09

(See the top two lines)

You can of course also verify this by querying Elasticsearch too for documents with `hash: ''`.

## Related issues

Closes https://github.com/elastic/kibana/issues/133230.

## Screenshots

An example of a partial result:

Screenshot 2022-06-07 at 17 42 56

It's important to note that the Logstash API doesn't send back `hash: ''`, it doesn't send the key at all, but the unmarshalling of the JSON will assign an empty string as the property is defined on the struct.

After the unmarshal we see something like this:

Screenshot 2022-06-07 at 18 02 24

## Logs

Screenshot 2022-06-12 at 00 51 09

### CODE REVIEW

File: metricbeat/module/logstash/node_stats/_meta/test/node_stats.710.json [hunk 1]
```diff
@@ -1,192 +1,197 @@
 {
-    "host": "anonymous",
-    "version": "7.8.1",
```
Comment (Kerry350) [inline metricbeat/module/logstash/node_stats/_meta/test/node_stats.710.json:3]: Regenerated this based off `7.10.0`.

**Review by matschaffer — COMMENTED** [review]
Makes general sense to me. Shame about all the retroactive-lint warnings. :(

The note I left on https://github.com/elastic/beats/pull/30896#issuecomment-1103519792 might help with the errors linting.

File: metricbeat/module/logstash/node_stats/data.go [hunk 2]
```diff
@@ -174,9 +176,20 @@ func eventMapping(r mb.ReporterV2, content []byte, isXpack bool) error {
...
+	// unit tests with much older fixture versions still work.
+	var PipelineDocumentsContainHashVersion = version.MustNew("7.3.0")
+	var StatsVersion = version.MustNew(nodeStats.Version)
+	pipelineDocumentsShouldContainHash := !StatsVersion.LessThan(PipelineDocumentsContainHashVersion)
+
 	for pipelineID, pipeline := range nodeStats.Pipelines {
-		pipeline.ID = pipelineID
-		pipelines = append(pipelines, pipeline)
+		if pipelineDocumentsShouldContainHash && (pipeline.Hash == "" || pipeline.Vertices == nil) {
+			logger.Warn("Pipeline document was discarded due to missing properties. This can happen when the Logstash node stats API is polled before the pipeline setup has completed.")
```
Comment (klacabane) [inline metricbeat/module/logstash/node_stats/data.go:188]: Would it be useful to include the pipelineID here ?

Comment (Kerry350) [reply metricbeat/module/logstash/node_stats/data.go:188]: It would, good point.

**Comment by Kerry350** [comment]
@matschaffer 

> Makes general sense to me. Shame about all the retroactive-lint warnings. :(
> The note I left on https://github.com/elastic/beats/pull/30896#issuecomment-1103519792 might help with the errors linting.

Thanks for the pointer. I'll take a look and see if it might actually be quick to get through them all. If not, I'll do it in a followup. Would be nice to get the UI going GA for 8.4.0 is all.

**Comment by Kerry350** [comment]
@matschaffer Going to merge as-is (well, with https://github.com/elastic/beats/pull/31868/commits/78f2dfab8d0e0ab4b71b9379cfd6c112c4979f10 added). And then come with a followup PR (I think that will also keep the changeset here clearer if anyone wants to check the behaviour change).

**Pull request merged**